proposals/0000-gadt-syntax.rst

```diff
@@ -0,0 +1,319 @@
+as "subject to change". We propose to fix a couple of issues associated with it
+and to advertise the resulting syntax as stable (that is, a breaking change to
+it would require another proposal).
+
+
+Motivation
+----------
+
+First and foremost, let us do away with the misconception that the GADT
+constructor signatures are type signatures.
+
+1. **Constructor-specific syntax**
+
+   There are syntactic forms that are not valid in type signatures but are
+   currently allowed in GADT constructor signatures::
+
+     data T a where
+       S :: !a -> T a
+       R :: { fld :: a } -> T a
+
+   In ``S``, the field is marked as strict using a prefix ``!``. In ``R``, the
+   field is given a name using record syntax ``{ ... }``. The signatures of these
+   constructors are perfectly fine, but they are rejected if used as function type
+   signatures::
+
+     s :: !a -> T a
+     s = S
+
+     r :: { fld :: a } -> T a
+     r = R
+
+   ``s`` and ``r`` are rejected with the following error messages::
+
+     GadtConSyn.hs:7:6: error:
+         Unexpected strictness annotation: !a
+         strictness annotation cannot appear nested inside a type
+
+     GadtConSyn.hs:10:6: error:
+         Record syntax is illegal here: {fld :: a}
+
+2. **Constructor-specific limitations**
+
+   There are also type signatures that are not valid GADT constructor
+   signatures::
+
+     k :: a -> Eq a => F
+
+   This is accepted (assuming a ``data F`` declaration elsewhere), but an
+   analogous constructor signature is not::
+
+     data F where
+       K :: a -> Eq a => F
+
+   ``K`` is rejected with the following error message::
+
+     GadtConSyn.hs:16:3: error:
+         GADT constructor type signature cannot contain
+         nested ‘forall’s or contexts
+
+So, if GADT constructor signatures are syntactically not type signatures, then
+what are they?  The User's Guide provides a BNF-style grammar for GADT
+constructor declarations, see `6.4.7.1. Formal syntax for GADTs
+<https://downloads.haskell.org/ghc/9.0.1/docs/html/users_guide/exts/gadt_syntax.html#formal-syntax-for-gadts>`_::
+
+  gadt_con ::= conids '::' opt_forall opt_ctxt gadt_body
+
+  conids ::= conid
+          |  conid ',' conids
+
+  opt_forall ::= <empty>
+              |  'forall' tv_bndrs '.'
+
+  tv_bndrs ::= <empty>
+            |  tv_bndr tv_bndrs
+
+  tv_bndr ::= tyvar
+           |  '(' tyvar '::' ctype ')'
+
+  opt_ctxt ::= <empty>
+            |  btype '=>'
+            |  '(' ctxt ')' '=>'
+
+  ctxt ::= ctype
+        |  ctype ',' ctxt
+
+  gadt_body ::= prefix_gadt_body
+             |  record_gadt_body
+
+  prefix_gadt_body ::= '(' prefix_gadt_body ')'
+                    |  return_type
+                    |  opt_unpack btype '->' prefix_gadt_body
+
+  record_gadt_body ::= '{' fieldtypes '}' '->' return_type
+
+  fieldtypes ::= <empty>
+              |  fieldnames '::' opt_unpack ctype
```

Comment:
@monoidal points out that I missed linearity syntax here.

Reply:
Actually, I’m not sure if we want the syntax from #111 or #370. They seem to differ.

* #111 gives this example:
   ```haskell
   data R where
      R :: { f1 %π :: A, f2 %ρ :: B } -> R
   ```

* #370 gives this grammar:
  ```haskell
  fielddecl ::= vars '::' (type | '!' atype) {modifier}
  ```

  Which would look as:

   ```haskell
   data R where
      R :: { f1 :: A %π, f2 :: B %ρ } -> R
   ```

Which one do we want? @monoidal @goldfirere

Reply:
I don't care. Maybe I didn't look at #111 closely enough when I wrote #370? Either syntax is compatible with how I see modifiers working in general: they affect the "next thing", if there is one. In the #111 syntax, the next thing is the `::`, which is a reasonable thing to affect. In the #370, there is no next thing, in which case the modifier affects the entire construct it appears in -- that is, the field declaration. So I don't much care. Thanks for noticing this discrepancy!

(If we go with the #111 syntax, I can change #370 accordingly.)

Reply:
This is pretty much my fault for not noticing the discrepancy when #370 was raised.

If nobody is strongly opposed, let's go with #111 's syntax: it has been examined more carefully. This was not a random proposition, but one that underwent several rounds of testing for least amount of surprise, and such (I may make it sound a bit more serious than it actually was, but it was still kind of seriously tested).

Reply:
Ok, I updated my proposal to use the #111 syntax.